server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java

```diff
@@ -100,7 +100,10 @@ protected Runnable cancelExistingRecoveryForBetterMatch(
         Metadata metadata = allocation.metadata();
         RoutingNodes routingNodes = allocation.routingNodes();
         ShardRouting primaryShard = allocation.routingNodes().activePrimary(shard.shardId());
-        assert primaryShard != null : "the replica shard can be allocated on at least one node, so there must be an active primary";
```

Comment:
@DaniilRoman do we know why this assertion had to be removed. @dblock In general I am not of fan of removing assertions, they exist in the system for a reason, we shouldn't let our guards down.

Reply:
@Bukhtawar the assertion has been replaced by an explicit null check, returning `null` instead of throwing an exception.  Either way the code will not proceed past this point. There is a previous `return null` at line 98, so it seems null is already an expected return value.

Reply:
@dbwiddis If I am understanding this correct there are two things
1. the method `cancelExistingRecoveryForBetterMatch` is expected to return a `Runnable` to denote some work to cancel existing recovery. A `null` as a method denotes no work is supposed to be done for cancelling. 
2. the invariant that says for a given shard(supposedly replica) it's corresponding `primary` should be non-null, since the allocator is supposed to kick in to assign a `replica` copy 
So there are two different things that exist and equating them IMO is incorrect. Removing the assertion essentially says at this point of execution a primary shard can be null when the system was designed ensuring that it shouldn't?

Reply:
> So there are two different things that exist and equating them IMO is incorrect. Removing the assertion essentially says at this point of execution a primary shard can be null when the system was designed ensuring that it shouldn't?

I haven't dug into the logic, but from the face of it, the previous situation is that the assertion causes an exception, and the exception isn't handled.

I've no objection to reverting this change if you think the exception should be caught and addressed elsewhere.

Reply:
>I haven't dug into the logic, but from the face of it, the previous situation is that the assertion causes an exception, and the exception isn't handled.

Unfortunately assertions don't run outside tests and the stack trace isn't an assertion tripping. Even if an assertion is failing the right thing to do is investigate why it is failing rather than burying the problem

Reply:
>I've no objection to reverting this change if you think the exception should be caught and addressed elsewhere.

Yes please let's revert this commit
cc: @dblock

Reply:
> Unfortunately assertions don't run outside tests

@Bukhtawar that's not the case with the `assert` keyword as it's used here.  It throws an `AssertionError`.

As an `Error` it's not normally caught by exception catch blocks unless you're specifically looking for it, and by the traditional interpretation of `Error` means that it's an unrecoverable situation and the entire JVM should be shut down.

Is it your opinion that in the case there is no primary shard, that the node should restart?

Reply:
@Bukhtawar @dblock  @SwethaGuptha @shiv0408 looking closer at the original issue (#13993) it produces an NPE at line 108, so it seems the stack trace on that issue is on a version prior to the one that added this `assert` statement.  The assertion was added in #10254 so it looks like the stack trace in #13993 was generated from the code prior to that other PR: https://github.com/opensearch-project/OpenSearch/blob/12115d1ad1feca522373220f5bef367f0c8008aa/server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java#L108 (CC:)

So generally my thoughts:
 - The NPE bug was already "fixed" by #13993 before the issue was filed
 - The "fix" replaced an NPE with an AssertionError instead which is changed behavior which might itself cause a node drop (except the issue indicates this is happening during a node drop)
 - I think we all need a bit more insight into when exactly this situation occurs and whether throwing an `Error` makes things worse

Reply:
@dbwiddis #10254 didn't "fix" or modify the behavior of this method, I had just refactored the core logic to a protected method which could be called from child batch allocator class. @SwethaGuptha can you please provide more details on which version of code you were running when you encountered this error.

Reply:
> #10254 didn't "fix" or modify the behavior of this method, I had just refactored ...

So I see looking again... sorry. The first time I was just looking for a line number.  But now I'm really confused.  In order for an NPE on line 108 on null `primaryShard` it would have had to pass the assertions on 106 and 107.
https://github.com/opensearch-project/OpenSearch/blob/12115d1ad1feca522373220f5bef367f0c8008aa/server/src/main/java/org/opensearch/gateway/ReplicaShardAllocator.java#L106-L108

Is this just undefined/unpredictable behavior during JVM shutdown (node drop)?

In any case, seems a revert of this PR is the right action at this point.

Reply:
> @Bukhtawar that's not the case with the assert keyword as it's used here. It throws an AssertionError.
As an Error it's not normally caught by exception catch blocks unless you're specifically looking for it, and by the traditional interpretation of Error means that it's an unrecoverable situation and the entire JVM should be shut down.

Sorry I don't think this is the correct understanding of how `assert` works. AssertionErrors are only configured for tests using a `-ea` JVM flag and in this case I don't think an `AssertionError` was triggered. Please feel free to read up more on this [here](https://docs.oracle.com/javase/8/docs/technotes/guides/language/assert.html#:~:text=An%20assertion%20is%20a%20statement,than%20the%20speed%20of%20light.)

>In any case, seems a revert of this PR is the right action at this point

Thanks for your understanding.

Reply:
Got it, and thanks for educating me.

So wouldn't an appropriate fix for the NPE then to both put the `assert` back _and_ null check to avoid the exception?  That seems odd but I've never really used assertions like this so at this point I really don't know the best approach.

Reply:
> So wouldn't an appropriate fix for the NPE then to both put the assert back and null check to avoid the exception?

@dbwiddis The appropriate fix is to figure out why the invariant is being violated and fix _that_ bug. i.e. if no primary exists, then there can be no recovery and this method should not be called. The previous behavior of having the code fail with an NPE is preferable to ignoring a situation that is not intended to be possible.

Slightly off topic, but in general I really dislike using the `assert` keyword. Instead of `assert primaryShard != null : "message"` I almost always prefer something like `Objects.requireNonNull(primaryShard, "message")` so that it deterministically fails in all situations. Only if the code is super performance critical and if the assertion check is computationally intensive do I think the `assert` keyword makes sense.

Reply:
> The appropriate fix is to figure out why the invariant is being violated and fix that bug. 

@andrross In general I agree. 

In reality, I read the issue reporting this bug and it indicates it's happening during a node drop.  Which means the JVM is shutting down for \<insert reason here\>.  Which means undefined/unpredictable behavior.

TLDR: I'm not sure there is a bug, and the symptom may be a result of a JVM being shut down causing a node drop as a result of some totally unrelated thing.

There's not enough information in the issue to infer more.